spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ValidationMethod.scala

```diff
@@ -219,7 +219,325 @@ class Top1Accuracy[T: ClassTag](
+    }
+
+    val confidenceArr = (0 until classes).map(_ => new ArrayBuffer[(Float, Boolean)]).toArray
+    if (_output.nDimension() == 2) {
+      (1 to _output.size(1)).foreach(i => {
+        val row = _output.select(1, i)
+        val gtClz = ev.toType[Float](_target.valueAt(i))
+        for(clz <- 0 until classes) {
+          confidenceArr(clz) += ((ev.toType[Float](row.valueAt(clz + 1)), gtClz == clz))
+        }
+      })
+    } else {
+      require(_output.dim() == 1, "The output should have 1 or 2 dimensions")
+      val row = _output
+      val gtClz = ev.toType[Float](_target.valueAt(1))
+      for(clz <- 0 until classes) {
+        confidenceArr(clz) += ((ev.toType[Float](row.valueAt(clz + 1)), gtClz == clz))
+      }
+    }
+    new MAPValidationResult(classes, k, confidenceArr, posCnt)
+  }
+
+  override def format(): String = s"MAP@$k"
+}
+
+object MAPUtil {
+
+  // find top k values & indices in a column of a matrix
+  def findTopK(k: Int, arr: Array[Array[Float]], column: Int): Array[(Int, Float)] = {
+    val q = collection.mutable.PriorityQueue[(Int, Float)]()(Ordering.by[(Int, Float), Float](_._2))
+    arr.indices.foreach(i => {
+      q.enqueue((i, arr(i)(column)))
+    })
+    val end = Math.min(k, q.size)
+    (1 to end).map(_ => q.dequeue()).toArray
+  }
+}
+
+/**
+ * The MAP Validation Result. The results are not calculated until result() or format() is called
+ * require class label beginning from 0
+ */
+class MAPValidationResult(
+  private val nClass: Int,
+  // take the first k samples, or -1 for all samples
+  private val k: Int,
+  // the predicts for each classes. (Confidence, GT)
+  private var predictForClass: Array[ArrayBuffer[(Float, Boolean)]],
+  private var gtCntForClass: Array[Int],
+  private val useVoc2007: Boolean = false,
+  private val skipClass: Int = -1
+)
+  extends ValidationResult {
+
+
+  if (skipClass < 0) {
+    require(skipClass == -1, s"Invalid skipClass $skipClass")
+  } else {
+    require(skipClass >= 0 && skipClass < nClass, s"Invalid skipClass $skipClass")
+  }
+
+  def calculateClassAP(clz: Int): Float = {
+    val posCnt = gtCntForClass
+    // for each class, first find top k confident samples
+    val sorted = predictForClass(clz).sortBy(v => v._1)(Ordering.Float.reverse) // decending order
+    var tp = 0
+    val refinedK = if (k > 0) k else sorted.size
+    // calculate the max precision for each different recall
+    // for each top-j items, calculate the (precision, recall)
+    val PnR = sorted.take(refinedK).zipWithIndex.flatMap { case (predict, j) =>
+      if (predict._2) {
+        // if it is a hit
+        tp += 1
+        // j + 1 is the total number of samples marked positive by the model
+        val precision = tp.toFloat / (j + 1)
+        val recall = tp.toFloat / posCnt(clz)
+        Iterator.single(recall, precision)
+      } else {
+        Iterator.empty
+      }
+    }
+
+    // get Average precision over each different recall
+    if (useVoc2007) {
```

Comment:
what does `useVoc2007` mean ?

Reply:
Pascal VOC uses different algorithms to calculate Average Precision before and after the year of 2007. It's a flag to indicate the algorithm.